File: src/ru/greeneyes/project/pomidoro/model/PomodoroModel.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ru.greeneyes.project.pomidoro.model;

import java.util.WeakHashMap;
import java.util.concurrent.TimeUnit;

import static ru.greeneyes.project.pomidoro.model.PomodoroModel.PomodoroState.BREAK;
import static ru.greeneyes.project.pomidoro.model.PomodoroModel.PomodoroState.RUN;
import static ru.greeneyes.project.pomidoro.model.PomodoroModel.PomodoroState.STOP;

/**
 * User: dima
 * Date: May 29, 2010
 */
public class PomodoroModel {
	private static final int PROGRESS_INTERVAL_MILLIS = 1000;

	public enum PomodoroState {
		STOP,
		RUN,
		BREAK
	}

	private final Settings settings;

	private PomodoroState state;
	private PomodoroState lastState;
	private long startTime;
	private int progressMax;
	private int progress;
	private boolean wasManuallyStopped;
	private final WeakHashMap<Object, Runnable> listeners = new WeakHashMap<Object, Runnable>();

	private final PomodoroModelState pomodoroModelState;

	public PomodoroModel(Settings settings, PomodoroModelState pomodoroModelState) {
		this.settings = settings;
		this.pomodoroModelState = pomodoroModelState;

		state = pomodoroModelState.state;
		lastState = pomodoroModelState.lastState;
		startTime = pomodoroModelState.startTime;

		if (pomodoroModelState.state == RUN) {
			long timeSincePomodoroStart = System.currentTimeMillis() - pomodoroModelState.startTime;
			boolean shouldNotContinuePomodoro = (timeSincePomodoroStart > settings.getTimeoutToContinuePomodoro());
			if (shouldNotContinuePomodoro) {
				state = STOP;
				lastState = null;
				startTime = -1;
			}
		}

		updateProgressMax();
		progress = progressMax;
	}

	public synchronized void switchToNextState() {
		switch (state) {
			case STOP:
				state = RUN;
				startTime = System.currentTimeMillis();
				updateProgressMax();
				break;
			case RUN:
				state = STOP;
				wasManuallyStopped = true;
				break;
			case BREAK:
				state = STOP;
				wasManuallyStopped = true;
				break;
			default:
				throw new IllegalStateException();
		}
		updateState();
	}

	public synchronized void updateState() {
		long time = System.currentTimeMillis();
		switch (state) {
			case RUN:
				updateProgress(time);
				if (time >= startTime + progressMax) {
					state = BREAK;
					startTime = time;
					updateProgress(time);
					updateProgressMax();
					settings.setPomodorosAmount(settings.getPomodorosAmount() + 1);
				}
				break;
			case BREAK:
				updateProgress(time);
				if (time >= startTime + progressMax) {
					state = STOP;
					wasManuallyStopped = false;
				}
				break;
			case STOP:
				if (lastState == STOP) {
					return;
				}
				break;
		}

		for (Runnable listener : listeners.values()) {
			listener.run();
		}

		if (lastState != state) {
			saveModelState();
		}
		lastState = state;
	}

	private void saveModelState() {
		pomodoroModelState.state = state;
		pomodoroModelState.lastState = lastState;
		pomodoroModelState.startTime = startTime;
	}

	public synchronized int getProgress() {
		return progress;
	}

	public synchronized int getProgressMax() {
		return progressMax / PROGRESS_INTERVAL_MILLIS;
	}

	public synchronized int getPomodorosAmount() {
		return settings.getPomodorosAmount();
	}

	public synchronized void resetPomodoros() {
		settings.setPomodorosAmount(0);
	}

	public synchronized PomodoroState getState() {
		return state;
	}

	public synchronized PomodoroState getLastState() {
		return lastState;
	}

	public synchronized boolean wasManuallyStopped() {
		return wasManuallyStopped;
	}

	public synchronized void addUpdateListener(Object key, Runnable runnable) {
		listeners.put(key, runnable);
	}

	private void updateProgress(long time) {
		progress = (int) ((time - startTime) / PROGRESS_INTERVAL_MILLIS);
		if (progress > getProgressMax()) {
			progress = getProgressMax();
		}
	}

	private void updateProgressMax() {
		switch (state) {
			case RUN:
				progressMax = (int) settings.getPomodoroLength();
				break;
			case BREAK:
				progressMax = (int) settings.getBreakLength();
				break;
		}
	}
}

File: src/pomodoro/model/PomodoroModel.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package pomodoro.model

import pomodoro.model.PomodoroModel.PomodoroState.*
import java.util.*


class PomodoroModel @JvmOverloads constructor(private val settings: Settings,
                                              private val pomodoroModelState: PomodoroModelState,
                                              now: Long = System.currentTimeMillis()) {

    enum class PomodoroState {
        /**
         * Pomodoro timer was not started or was stopped during pomodoro or break.
         */
        STOP,
        /**
         * Pomodoro in progress.
         */
        RUN,
        /**
         * Pomodoro during break. Can only happen after completing a pomodoro.
         */
        BREAK
    }

    @get:Synchronized var state: PomodoroState? = null
        private set
    @get:Synchronized var lastState: PomodoroState? = null
        private set
    private var startTime: Long = 0
    private var progressMax: Int = 0
    @get:Synchronized var progress: Int = 0
        private set
    @get:Synchronized var pomodorosAmount: Int = 0
        private set
    private var wasManuallyStopped: Boolean = false
    /**
     * It's a WeakHashMap to make it simpler to automatically remove listeners.
     * The most common usage is when there are several IntelliJ windows, UI components subscribe to model and
     * then window is being closed.
     */
    private val listeners = WeakHashMap<Any, () -> Unit>()

    init {

        loadModelState(now)

        updateProgressMax()
        progress = progressMax
    }

    @Synchronized fun onUserSwitchToNextState() {
        onUserSwitchToNextState(System.currentTimeMillis())
    }

    @Synchronized fun onUserSwitchToNextState(time: Long) {
        when (state) {
            STOP -> {
                state = RUN
                startTime = time
                updateProgressMax()
            }
            RUN -> {
                state = STOP
                wasManuallyStopped = true
            }
            BREAK -> {
                state = STOP
                wasManuallyStopped = true
            }
            else -> throw IllegalStateException()
        }
        onTimer(time)
    }

    @Synchronized fun onTimer() {
        onTimer(System.currentTimeMillis())
    }

    @Synchronized fun onTimer(time: Long) {
        when (state) {
            RUN -> {
                updateProgress(time)
                if (time >= startTime + progressMax) {
                    state = BREAK
                    startTime = time
                    updateProgress(time)
                    updateProgressMax()
                    pomodorosAmount++
                }
            }
            BREAK -> {
                updateProgress(time)
                if (time >= startTime + progressMax) {
                    state = STOP
                    wasManuallyStopped = false
                }
            }
            STOP -> if (lastState == STOP) {
                return
            }
        }

        for (listener in listeners.values) {
            listener.invoke()
        }

        if (lastState != state) {
            lastState = state
            saveModelState(time)
        }
        lastState = state
    }

    @Synchronized fun getProgressMax(): Int {
        return progressMax / PROGRESS_INTERVAL_MILLIS
    }

    @Synchronized fun resetPomodoros() {
        pomodorosAmount = 0
        pomodoroModelState.pomodorosAmount = pomodorosAmount
    }

    @Synchronized fun wasManuallyStopped(): Boolean {
        return wasManuallyStopped
    }

    @Synchronized fun addUpdateListener(key: Any, runnable: () -> Unit) {
        listeners.put(key, runnable)
    }

    private fun loadModelState(now: Long) {
        state = pomodoroModelState.pomodoroState
        lastState = pomodoroModelState.lastState
        startTime = pomodoroModelState.startTime
        pomodorosAmount = pomodoroModelState.pomodorosAmount

        if (pomodoroModelState.pomodoroState != STOP) {
            val timeSincePomodoroStart = now - pomodoroModelState.lastUpdateTime
            val shouldNotContinuePomodoro = timeSincePomodoroStart > settings.timeoutToContinuePomodoro
            if (shouldNotContinuePomodoro) {
                state = STOP
                lastState = null
                startTime = -1
                saveModelState(now)
            }
        }
    }

    private fun saveModelState(now: Long) {
        pomodoroModelState.pomodoroState = state
        pomodoroModelState.lastState = lastState
        pomodoroModelState.startTime = startTime
        pomodoroModelState.lastUpdateTime = now
        pomodoroModelState.pomodorosAmount = pomodorosAmount
    }

    private fun updateProgress(time: Long) {
        progress = ((time - startTime) / PROGRESS_INTERVAL_MILLIS).toInt()
        if (progress > getProgressMax()) {
            progress = getProgressMax()
        }
    }

    private fun updateProgressMax() {
        when (state) {
            RUN -> progressMax = settings.pomodoroLengthInMillis.toInt()
            BREAK -> progressMax = settings.breakLengthInMillis.toInt()
        }
    }

    companion object {
        private val PROGRESS_INTERVAL_MILLIS = 1000
    }
}